### PR TITLE
Improve GUI error dialog when no runtime is installed

### PR DESCRIPTION
**Pull request by mateoatr**
This adds information about the runtime version and architecture that the user needs to install in order to execute the failing application. Related to #3816.

Here's an example of how the dialog would look like: 
image

This makes the implicit assumption that, because we are running a GUI app, the user needs to install desktop runtime.

### CODE REVIEW

File: src/installer/tests/HostActivation.Tests/PortableAppActivation.cs [hunk 1]
```diff
@@ -554,6 +554,40 @@ public void AppHost_GUI_FrameworkDependent_MissingRuntimeFramework_ErrorReported
             }
         }
 
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void AppHost_GUI_RuntimeMissing_PrintArchitectureAndRuntimeVersionOnDialog()
```
Comment (elinor-fung) [inline src/installer/tests/HostActivation.Tests/PortableAppActivation.cs:559]: > PrintArchitectureAndRuntimeVersionOnDialog

The test isn't actually checking what is on the dialog, right? Just checking that a dialog is shown and that we log about showing it and the URL.

File: src/installer/tests/HostActivation.Tests/PortableAppActivation.cs [hunk 2]
```diff
@@ -554,6 +554,40 @@ public void AppHost_GUI_FrameworkDependent_MissingRuntimeFramework_ErrorReported
...
+                .Copy();
+
+            string appExe = fixture.TestProject.AppExe;
+            File.Copy(sharedTestState.BuiltAppHost, appExe, overwrite: true);
+            AppHostExtensions.BindAppHost(appExe);
+            AppHostExtensions.SetWindowsGraphicalUserInterfaceBit(appExe);
+
+            string invalidDotNet = SharedFramework.CalculateUniqueTestDirectory(Path.Combine(TestArtifact.TestArtifactsPath, "guiErrors"));
+            Directory.CreateDirectory(invalidDotNet);
+            File.Copy(sharedTestState.BuiltDotNet.DotnetExecutablePath, Path.Combine(invalidDotNet, "dotnet.exe"), true);
```
Comment (elinor-fung) [inline src/installer/tests/HostActivation.Tests/PortableAppActivation.cs:571]: Why do we need to copy over `dotnet`? Is just creating the directory not enough?

File: src/installer/tests/HostActivation.Tests/PortableAppActivation.cs [hunk 3]
```diff
@@ -554,6 +554,40 @@ public void AppHost_GUI_FrameworkDependent_MissingRuntimeFramework_ErrorReported
...
+            WaitForPopupFromProcess(command.Process);
+            command.Process.Kill();
+
+            var expectedErrorCode = Constants.ErrorCode.CoreHostLibMissingFailure.ToString("x");
+            var result = command.WaitForExit(true)
+                .Should().Fail()
+                .And.HaveStdErrContaining($"Showing error dialog for application: '{Path.GetFileName(appExe)}' - error code: 0x{expectedErrorCode}")
+                .And.HaveStdErrContaining($"url: 'https://aka.ms/dotnet-core-applaunch?missing_runtime=true")
+                .And.HaveStdErrContaining("gui=true")
+                .And.HaveStdErrContaining($"&apphost_version={sharedTestState.RepoDirectories.MicrosoftNETCoreAppVersion}");
```
Comment (elinor-fung) [inline src/installer/tests/HostActivation.Tests/PortableAppActivation.cs:588]: This test is largely overlapping with AppHost_GUI_FrameworkDependent_MissingRuntimeFramework_ErrorReportedInDialog above where missingHostfxr=true. I think it would make sense to just remove the missingHostfxr=true case from the test above so that it is just the test for a missing framework (NETCoreApp) and this test can be the one for missing runtime (hostfxr).

File: src/installer/tests/HostActivation.Tests/PortableAppActivation.cs (hunk 1 above)
Comment (mateoatr) [reply src/installer/tests/HostActivation.Tests/PortableAppActivation.cs:559]: Correct. AFAIK we currently don't have any tests checking the actual text on the dialog. ~Renaming.

File: src/installer/tests/HostActivation.Tests/PortableAppActivation.cs (hunk 2 above)
Comment (mateoatr) [reply src/installer/tests/HostActivation.Tests/PortableAppActivation.cs:571]: No need.

**Review by elinor-fung — APPROVED** [review]

File: src/installer/tests/HostActivation.Tests/PortableAppActivation.cs (hunk 3 above)
Comment (mateoatr) [reply src/installer/tests/HostActivation.Tests/PortableAppActivation.cs]: Split above's test into AppHost_GUI_FrameworkDependent_MissingRuntimeFramework_ErrorReportedInDialog and AppHost_GUI_MissingRuntime_ErrorReportedInDialog. The two are pretty much the same only difference are the URL that should be produced and the error code -- I think having the two would be more valuable if we had a way to check the dialog text.

File: src/native/corehost/apphost/apphost.windows.cpp [hunk 4]
```diff
@@ -81,6 +82,7 @@ namespace
         {
             // We don't have a great way of passing out different kinds of detailed error info across components, so
             // just match the expected error string. See fx_resolver.messages.cpp.
+            dialogMsg = pal::string_t(_X("To run this application, you must install .NET.\n\n"));
```
Comment (vitek-karas) [inline src/native/corehost/apphost/apphost.windows.cpp:85]: I would probably reword this to something like: "To run this application, you must install missing framework for .NET." - The "missing .NET" case is handled above, this is the "missing framework case" (and we MUST NOT use the ".NET framework" wording - since that would make it sound like the .NET Framework product.

File: src/native/corehost/apphost/apphost.windows.cpp [hunk 5]
```diff
@@ -61,11 +61,12 @@ namespace
         if (pal::getenv(_X("DOTNET_DISABLE_GUI_ERRORS"), &gui_errors_disabled) && pal::xtoi(gui_errors_disabled.c_str()) == 1)
             return;
 
-        pal::string_t dialogMsg = _X("To run this application, you must install .NET.\n\n");
+        pal::string_t dialogMsg;
         pal::string_t url;
         const pal::string_t url_prefix = _X("  - ") DOTNET_CORE_APPLAUNCH_URL _X("?");
         if (error_code == StatusCode::CoreHostLibMissingFailure)
         {
+            dialogMsg = pal::string_t(_X("To run this application, you must install .NET Desktop Runtime ")) + _STRINGIFY(COMMON_HOST_PKG_VER) + _X(" (") + get_arch() + _X(").\n\n");
```
Comment (vitek-karas) [inline src/native/corehost/apphost/apphost.windows.cpp:69]: Thanks for adding the architecture to the error. I think we should also add this to the missing framework error.
There are multiple places for this, but the most important ones are:
https://github.com/dotnet/runtime/blob/e6e7ef94b78bd73e84fc4c58138535d6e5b39911/src/native/corehost/fxr/fx_resolver.messages.cpp#L115
and
https://github.com/dotnet/runtime/blob/e6e7ef94b78bd73e84fc4c58138535d6e5b39911/src/native/corehost/fxr/fx_resolver.messages.cpp#L119

I would just add "(x86)" to the message after the framework name (or version).

It's a minor change, but I think this will be crucial for the ARM64 cases.
